NFC for SR-12085 Clean up TypeCheckType so it never returns Type().

dfsweeney opened the pull request:
Added ErrorType::get(Context) to TypeCheckType.cpp and clients.
Replaced assertions checking for null pointers with type->hasError() checks.
Note that getDynamicBridgedThroughObjCClass can still return null pointers so
the assert checks are still there for callers.

<!-- What's in this pull request? -->
No functional change. This cleans up some instances of return Type() in TypeCheckType.cpp. There are a lot of other instances of return Type() but this gets the 15 of them in TypeCheckType.cpp plus a few others in direct clients to the functions in TypeCheckType.cpp that returned Type(). @CodaFi helped me with this. 

This is my first one so I'm not sure I'm doing it right. Running all the validation tests gives me:

    Expected Passes    : 6490
    Expected Failures  : 5
    Unsupported Tests  : 100
    Unexpected Failures: 205

both before and after this change when I build with ninja. A lot of the validation test failures are in stdlib--I am not sure why.

The two validation test failures in Sema are:

    Swift(macosx-x86_64) :: Sema/type_checker_perf/fast/rdar42672946.swift
    Swift(macosx-x86_64) :: Sema/type_checker_perf/slow/rdar23327871.swift.gyb
   
I checked out swift-DEVELOPMENT-SNAPSHOT-2020-02-08-a and then put the commit directly on top of that. I hope that is OK--I tried a couple of things but that seemed to build clean. 

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves SR-12085.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

owenv commented:
@dfsweeney Welcome to the project! Depending on your build setup, it's not unusual for many of the validation tests to fail locally trying to import `StdlibCollectionUnitTest`, which may be the case here. Those two `Sema` tests are both performance tests, so it's possible those are local-only failures too, but I'm less sure. 

Let's try running tests on CI and see what happens.

owenv commented:
@swift-ci please smoke test

dfsweeney commented:
I found a bug I introduced that caused the smoke test to fail in 4 tests. Note I still have one failure in tests, in Parse/operators.swift. I'm working on that one. A lot of the validation tests are failing on import StdlibCollectionUnittest where the module is not found. I probably need to add something to the build-script to figure that out.

dfsweeney commented:
Cleared up one of the errors in Parse/operators.swift. Still getting

     /Users/dfs/swift-source/swift/test/Parse/operators.swift:125:5: error: unexpected error produced: broken standard library: cannot find intrinsic operations on Optional<T> foo??bar // expected-error{{broken standard library}} expected-error{{consecutive statements}} {{6-6=;}}

The pattern of returning Type() and then checking the result for null is pretty common. I thought I got all the clients of the original 15 in TypeCheckType but they propagate up and down the call chain. In general replacing 

  if (!type) 

with 

  if (!type || type->hasError()) 

keeps the logic in the same flow, but it's pretty easy to miss that and then let the execution run into code that does other things. Sometimes you get a diagnostic that triggers a test failure, but I'm not sure there's 100% coverage. 

If I can figure out where I missed a check that's causing the one remaining test failure the patch would run clean, but I'm a little worried that this might cause more problems at a distance.

CodaFi commented:
You can assert as a postcondition along the APIs you convert then. Asserting in the clients is gonna be a game of whack-a-mole, as you’ve already discovered.

dfsweeney commented:
I cleared the last warning in the tests. 

@swift-ci please smoke test

dfsweeney commented:
(I'm not sure if I can trigger a smoke test myself, if not can one of you please start it? Thanks!)

theblixguy commented:
Only people with commit access can.

@swift-ci please test

dfsweeney commented:
Thanks!

dfsweeney commented:
I rebased these on top of the swift-5.2-branch locally. (There was one merge conflict that I had to clear up. it builds and all the tests pass.) Can I push that to the branch for this pull request or will things really get weird? I think it would change the base branch for the pull request and I don't know if that's OK or not in GitHub land. Any ideas?

CodaFi commented:
> I rebased these on top of the swift-5.2-branch locally. 

You shouldn't rebase on top of a release branch. Master is where this patch is going to land, and if we need to cherry-pick it, we'll do it from there. 5.2 is missing a whole heackuva lot of commits.

CodaFi commented on lib/Sema/TypeCheckType.cpp:
```diff
@@ -382,35 +382,35 @@ Type TypeChecker::getStringType(ASTContext &Context) {
     return typeDecl->getDeclaredInterfaceType();
 
-  return Type();
+  return ErrorType::get(Context);
```
`ErrorType`'s contract is that if one has been created, a diagnostic *must* have been emitted either at the point of creation or upstream of it. We have a set of arbitrary diagnostics for when internal utilities like this fail (e.g. `diag::sugar_type_not_found`). We should emit something here and in the others.

dfsweeney replied:
Ok, add a diagnostic emit for all 15 return ErrorType(ctx) in TypeCheckType.cpp and any others I added.

dfsweeney replied:
For getStringType, getSubstringType, getIntType, getInt8Type, getUInt8Type, sugar_type_not_found is not really appropriate right? Since they're not really syntactic sugar for anything. It looks like these guys get called for equality checks mainly. If the underlying method comes from KNOWN_STDLIB_TYPE_DECL() in KnownStdlibTypes.def is it possible for these to fail at all? Or only if the stdlib is really broken/missing? I may have answered my question there by thinking out loud. Should I add a new diagnostic for 'broken standard library' with the associated types? or add more select cases to the existing sugar_type_not_found diagnostic?

dfsweeney replied:
I started on adding to the existing sugar_type_not_found diagnostic, but emitting a diagnostic requires a source location, and these methods (and their clients) don't have one. I worked up the call chain a bit and could not find a useful source loc in the client methods (I was not exhaustive.) These getXType methods for stdlib types are kind of unusual but I don't completely understand why. I'll keep looking at it.

dfsweeney replied:
I put llvm_unreachable() in the cases where there's a broken standard library. That might be OK but it would still be better to emit a diagnostic although I'm having mechanical problems emitting one in that case. 

TypeChecker::getDynamicBridgedThroughObjCClass returns ErrorType if bridging fails but TypeChecker::TypeCheckerTypedCast continues and can either return a type or generate a diagnostic. Arguably that's OK. Alternately returning ErrorType in that situation is not appropriate and there should be a UnmatchedType kind of thing to indicate that situation. 

Still need diagnostics for TypeChecker::resolveOpaqueReturnType I think. 

@CodaFi do you have suggestions or opinions about this? The above is reflected in commit 
02d34f9

CodaFi commented on lib/Sema/TypeCheckType.cpp:
```diff
@@ -2634,7 +2633,7 @@ Type TypeResolver::resolveASTFunctionType(
                                    repr->getGenericEnvironment() != nullptr,
                                    diffKind, params)) {
-    return Type();
+    return ErrorType::get(Context);
```
`resolveASTFunctionTypeParams` returns `Type()` if `resolveType` returns `Type()`. But after your changes, `resolveType` should never return `Null`, and you'll automatically have an error type in the parameter list, and that parameter list can just be used directly. You can refactor `resolveASTFunctionTypeParams` to return the parameters directly now.

dfsweeney replied:
For this one, should resolveASTFunctionTypeParams return params instead of bool?

dfsweeney replied:
params is a SmallVectorImpl<>. it looks like llvm wants that to be an out parameter not a return value (based on the documentation and on that it does not have a copy constructor). If I rewrite resolve to return a SmallVectorImpl it does not like it, unless I return a pointer, which it then has to delete, etc. Maybe just call resolveASTFunctionTypeParams and remove the check? I can replace     if (!ty) return true; with an assertion in resolveASTFunctionTypeParams. It would always return false then.

CodaFi replied:
Sure. So we can change the return type to a suitably-sized SmallVector of parameters.

dfsweeney replied:
I just pushed a patch that changes the return type. I missed that SmallVector was the thing you return and that SmallVectorImpl can be an out parameter.

CodaFi commented on lib/Sema/TypeCheckType.cpp:
```diff
@@ -3159,9 +3158,7 @@ Type TypeResolver::resolveArrayType(ArrayTypeRepr *repr,
-  if (!sliceTy)
-    return ErrorType::get(Context);
-
+  assert(sliceTy && "TypeChecker::getArraySliceType returned nullptr.");
```
Once you're done here, please remove these asserts. The value being non-null is a structural part of the API now.

dfsweeney replied:
Should I delete all the null-check asserts that I added after testing and before pushing? I noticed that they should never fire unless I missed an upstream fix.

CodaFi replied:
Sure. They're a good refactoring signal, but they're going to become noise once this is all finished.

CodaFi commented on lib/Sema/TypeCheckType.cpp:
```diff
@@ -3725,13 +3720,13 @@ Type swift::resolveCustomAttrType(CustomAttr *attr, DeclContext *dc,
   ASTContext &ctx = dc->getASTContext();
   if (TypeChecker::validateType(ctx, attr->getTypeLoc(), resolution, options))
-    return Type();
+    return ErrorType::get(ctx);
```
The type returned from `validateType` should never be null. You can just propagate the type in the `TypeLoc` here. This'll help us kill off `TypeLoc` as a construct eventually.

dfsweeney replied:
return attr->getTypeLoc().getType() instead in the check on validateType(). We still need the check though right? It looks like it might be safe to just call validateType and continue but I'm not 100% because I may be missing some mechanics.

CodaFi replied:
`validateType` is just returning whether the type in the typeloc has an error type in it anyways. You should still early-exit, just early-exit with `attr->getTypeLoc().getType()` instead of a new error type.

dfsweeney replied:
I just pushed a patch that addresses this one.

dfsweeney commented:
Thanks for reviewing! I replied to each of the above, I'm pretty sure I understand but it helps to write it out. Sorry for the message generation. For the basing/rebasing, should I rebase this on to master and push it to the branch associated with the pull request? I was having some trouble getting the master branch to build. 

I will work through the changes above.

CodaFi commented:
> For the basing/rebasing, should I rebase this on to master and push it to the branch associated with the pull request?

You shouldn't need to keep rebasing onto master unless there's merge conflicts you have to resolve. But if you need to, pull and rebase onto master. I usually set a remote called `upstream` that points to Apple's copy of swift and run

```
$ git pull --rebase upstream master
```
every so often to keep in sync.

> I was having some trouble getting the master branch to build.

Do you have any logs or errors to share?

dfsweeney commented:
Thanks again! wrt the build problems, I was not using the swift/master branch of llvm, which made a lot of stuff flip out. At some point I just checked out the master branch (without swift/). Just one of those things getting used to a project, finding things you can break by accident. I am doing something like what you have above with an upstream branch. I am pushing through the rest of the above.

dfsweeney commented:
ok that's not what I wanted to happen, I think I left something out of the git push.

CodaFi reviewed (changes requested):
My high-level comment remains the same: The asserts about structural properties need to come out. If they cannot be removed with confidence, perhaps we should break this PR down into smaller pieces and work from the leaf functions up incrementally.

dfsweeney commented:
I cleared the structural assertions and pushed up but something looks weird about this--now it says that I want to merge 945 commits, which looks wrong. Anyway, can somebody fire off the checks at least? I will try to figure out what I did or did not do. If necessary I can close this PR and open a new one but then the history will be lost.

CodaFi commented:
I believe you pulled without a rebase. Try 
```
git pull --rebase upstream master
```

Where `upstream` is a remote entry pointing to apple/swift

dfsweeney commented:
I thought I rebased before I pushed but I guess not—maybe I was looking away. I think I rebased correctly this time but there were some merge conflicts that kept coming back, and the commits are out of order. Maybe I should cherry-pick it instead. I pushed the branch up with a force but I’m still not sure it’s right. I’m going to build and run the tests locally and see if I can figure it out. 

Thanks,

Dan

> On Mar 11, 2020, at 5:04 PM, Robert Widmann <notifications@github.com> wrote:
> 
> 
> I believe you pulled without a rebase. Try
> 
> git pull --rebase upstream master
> Where upstream is a remote entry pointing to apple/swift
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub <https://github.com/apple/swift/pull/29780#issuecomment-597922767>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAAWZBCDX4IY7LMQEM2AS4TRHAKGNANCNFSM4KTSIRWQ>.
>

dfsweeney commented:
I need to clean this up--somehow I rebased the same commits on top of master more than once.

dfsweeney commented:
I rebased and then squashed all the commits for this PR down to one, and force-pushed it to the branch attached to this PR. I think that is the thing that every piece of documentation for git-rebase says not to do of course. Or I might have done it wrong. It says 15 commits but should really just be one. Can we fire off the tests on this one?

CodaFi commented:
@swift-ci test

CodaFi commented on lib/Sema/CSGen.cpp:
```diff
@@ -1743,7 +1743,7 @@ namespace {
 
       Type optTy = getOptionalType(expr->getSubExpr()->getLoc(), valueTy);
-      if (!optTy)
+      if (!optTy || optTy->hasError())
```
This should never be null.

CodaFi commented on lib/Sema/CSGen.cpp:
```diff
@@ -3197,7 +3198,7 @@ namespace {
 
       Type optTy = getOptionalType(expr->getSubExpr()->getLoc(), valueTy);
-      if (!optTy)
+      if (!optTy || optTy->hasError())
```
This should never be null

dfsweeney replied:
I wonder why I left the two checks in. Probably an oversight but I will check whether I had noticed some case where optTy could be nullptr.

CodaFi commented on lib/Sema/TypeCheckType.cpp:
```diff
@@ -382,48 +382,50 @@ Type TypeChecker::getStringType(ASTContext &Context) {
                                                Type valueType) {
   // We can only bridge from class or Objective-C existential types.
+  // if we return ErrorType here, callers will continue trying to type check
+  // and return diagnostic if it fails.
```
This is not an invariant we have control over from this interface. If we can't diagnose here, we shouldn't make this transformation unless we update the callers to be more defensive.

dfsweeney replied:
The original logic had TypeChecker::getDynamicBridgedThroughObjCClass returning Type() if the bridge failed, and the client (repairFailures) continuing to try to match the type. Returning ErrorType is a little clunky because it's really indicating a type match failure in this leaf of the tree. It does not return a diagnostic (violating the contract of ErrorType) because it's not a diagnostic-worthy error yet. If the whole matchType->repairFailures process fails then you get a diagnostic. That was what I meant in the comment but it's not that clear. If there were a 'NoMatchType' that we could return for this situation it might be clearer but that's also a wider change and adds state to the system.

CodaFi commented:
There's some good cleanups in here. If we chunk this apart, I think we can merge them in this order:
- The Parameter cleanup in TypeCheckType
- The known stdlib decls bits

The rest require more infrastructure be converted first - I don't feel comfortable doing this completely piecemeal. We should work from the leaves (like the known stdlib decls parts) up.

dfsweeney commented:
I agree with you, there's a lot of infrastructure going on. I will leave in the Parameter cleanup and the changes that don't depend on client behavior and take out the changes that rely too much on what the callers do, if that makes sense.

dfsweeney committed:
NFC for SR-12085 Clean up TypeCheckType so it never returns Type().

Replaced "return Type();" in TypeChecker::getXType() methods in
TypeCheckType.cpp with "return ErrorType.get(ctx);" along with
appropriate diagnostics. Type() is equivalent to nullptr. Cleaned up
clients to test for Type->hasError rather than nullptr.

dfsweeney commented:
I rebuilt this with git reset/git add -p and force-updated the branch. This should have only the straightforward changes and not the other stuff we've been talking about. @CodaFi please let me know if I left anything in there that you don't expect or vice versa. Thanks!

CodaFi commented:
@swift-ci test

dfsweeney commented:
This looks like an infrastructure failure although those are the most famous of last words. Can we restart the test or see what happened? Part of the build from Hudson is below: 

—Dan

PASS: Swift(iphonesimulator-i386) :: Interpreter/dynamic_replacement_without_previous_calls.swift (11077 of 13130)
PASS: Swift(iphonesimulator-i386) :: Interpreter/dynamic_cast_set_conditional_type_metadata.swift (11078 of 13130)
FATAL: command execution failed
java.io.EOFException
	at java.io.ObjectInputStream$PeekInputStream.readFully(ObjectInputStream.java:2735)
	at java.io.ObjectInputStream$BlockDataInputStream.readShort(ObjectInputStream.java:3210)
	at java.io.ObjectInputStream.readStreamHeader(ObjectInputStream.java:895)
	at java.io.ObjectInputStream.<init>(ObjectInputStream.java:357)
	at hudson.remoting.ObjectInputStreamEx.<init>(ObjectInputStreamEx.java:49)
	at hudson.remoting.Command.readFrom(Command.java:140)
	at hudson.remoting.Command.readFrom(Command.java:126)
	at hudson.remoting.AbstractSynchronousByteArrayCommandTransport.read(AbstractSynchronousByteArrayCommandTransport.java:36)
	at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:63)
Caused: java.io.IOException: Unexpected termination of the channel
	at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:77)
Caused: java.io.IOException: Backing channel 'macOS-60' is disconnected.
	at hudson.remoting.RemoteInvocationHandler.channelOrFail(RemoteInvocationHandler.java:214)
	at hudson.remoting.RemoteInvocationHandler.invoke(RemoteInvocationHandler.java:283)
	at com.sun.proxy.$Proxy73.isAlive(Unknown Source)


> On Apr 22, 2020, at 6:39 PM, swift-ci <notifications@github.com> wrote:
> 
> 
> Build failed
> Swift Test OS X Platform <https://ci.swift.org/job/swift-PR-osx/20082/>
> Git Sha - 38c3cf3 <https://github.com/apple/swift/commit/38c3cf350cc89d4de5e5685c441e574ded81d281>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub <https://github.com/apple/swift/pull/29780#issuecomment-618110612>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAAWZBCPCCLTM7T66MRZFM3RN6E3HANCNFSM4KTSIRWQ>.
>

CodaFi commented:
Builder died

@swift-ci test macOS

varungandhi-apple commented:
@CodaFi is this good to merge?

varungandhi-apple commented:
Oof, there's a merge conflict now @dfsweeney 😬

dfsweeney commented:
I’ll take a look at this one again. 

—Dan

> On Apr 30, 2020, at 7:20 PM, Varun Gandhi <notifications@github.com> wrote:
> 
> 
> Oof, there's a merge conflict now @dfsweeney <https://github.com/dfsweeney> 😬
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/apple/swift/pull/29780#issuecomment-622202709>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAAWZBDALVZJ4VZHHSM6MTDRPIPUVANCNFSM4KTSIRWQ>.
>

CodaFi commented:
@dfsweeney Apologies for leaving this by the wayside - and for causing all of these merge conflicts. I believe I'm done messing around with type resolution for the time being. If you'd like, I can cherry-pick and shepherd this patch through myself and have you review the changes.

CodaFi commented:
To get some context down: I'm in the process of converting the mutating usages of `TypeLoc` into either requests or just breaking apart the `TypeLoc`. I was hoping we could merge this patch as a capstone to that refactoring process, but there have been intervening issues in TypeCheckType that required my immediate attention lately.

dfsweeney commented:
Hi @CodaFi it's all good, a lot of things are going on. 

I just force-pushed again to clear the merge conflict. There are a couple of llvm_unreachables in TypeCheckType.cpp that I have been hoping I would think of a better solution for. 

If you could cherry-pick it through that would be great! Let me know what you need me to do. Thanks!

CodaFi commented:
@swift-ci test

dfsweeney commented:
I fixed that one spot where we crossed wires on the ASTContext parameter from TypeChecker::validateType.

theblixguy commented:
@swift-ci please test

CodaFi commented:
Hey @dfsweeney, I finally cleared the blocker to doing this refactoring. I've taken this patch as the basis for a cleanup in https://github.com/apple/swift/pull/32321. I'm gonna close this out. Please feel free to review the linked PR if you have some time.

dfsweeney commented:
Thanks! I'm glad this finally came in for a landing.
